src/models/room-state.ts

```diff
@@ -420,6 +420,12 @@ export class RoomState extends TypedEventEmitter<EmittedEvents, EventHandlerMap>
             }
 
             const lastStateEvent = this.getStateEventMatching(event);
+            // Safety measurs to not update the room (and emit the update) with older state.
+            // The sync loop really should not send old events but it does very regularly.
+            // Logging on return in those two conditions results in a large amount of logging. (on startup and when running element)
+            if ((lastStateEvent?.event.origin_server_ts ?? 0) > (event.event.origin_server_ts ?? 1)) return;
```

Comment:
That is why I am labeling this as a workaround.
> This is a workaround at best and mostly serves the purpose to show that this issue exist and to have a starting point to look into the underlying problem.

There is also this chat (monologue) for context https://matrix.to/#/!jxlRxnrZCsjpjDubDX:matrix.org/$2X7ysozWciiyDHVev8iJ5yaFcZbZr57H2u8v5-7ysXw?via=matrix.org&via=envs.net&via=mozilla.org

>This is not how we do state resolution however...
(using the timestamp) This logic would need to use the replaces_state to be correct... But there is no good condition at which I can stop paginating back to see if there is really no previous state.

Reply:
Sure, but its a workaround which will cause the state to be wrong in the other direction whenever someone has the time set wrong on their server, which can also be done maliciously.

Reply:
I 100% agree that the timestamp fix is bad.
I really hope, that we dont merge any of this tbh.
It is just very clear in this PR what situation the current sync causes and where the issue lies.
Since the replces_state condition also covers the acute issue with calls being closed on join this could be a hotfix for the underlying issue that we can even reach this function with old state.

Do you have any idea why we do get some events twice and in the wrong order?

Reply:
Also this code would brake for maintaining `oldState` on a room, as the state there does indeed go reverse and backwards in time

> Do you have any idea why we do get some events twice and in the wrong order?

Probably due to the above, handling the edge in two directions, at each end of the live timeline

Reply:
But when i read the state I am using `getLiveTimeline().getState(FORWARD)` so we should always read from  the correct state.

Reply:
Sure, but this method you are editing is in both Forward & Backward RoomState classes. So the timestamp comparison would at least need inverting for the Backward case.

Reply:
> Where do you think one should look for a proper fix?

Would have to try trace what is causing events to be emitted out of order, I'd suggest `console.trace` on a dev build on which you try reproduce the issue

Reply:
Since this is a very fundamental issue I am wondering if it if someone who has long experience with the sync code is a better fit to investigate this?
Putting a breakpoint at the return in this pr allows to reproduce it.

I think I will try to adjust the pr like described here:
>  I think i will change this pr to not do any timestamp logic but only use the replaces_state condition and only in the case we are adding to endState.

so it is actually a valid hotfix.

Reply:
I updated this to respect `toStartOfTimeline` and removed the timestamp check.

Reply:
I am also wondering why we don't do any ordering based on `toStartOfTimeline`.
`setStateEvents` should do `stateEvents.reverse().forEach` if we insert at the start of the timeline (`startState.setStateEvents`)

Reply:
> Since this is a very fundamental issue I am wondering if it if someone who has long experience with the sync code is a better fit to investigate this?


I'm unfortunately not the best person for this

> so it is actually a valid duct tape fix.

Agreed

It'd be good to isolate the changes in this PR those those concerning the title of the PR, the RTC stuff is unrelated. It'd also be good to have tests to assert this does the right thing

Reply:
Oh definitely. not exactly sure how those changes sneaked in...